### DISCUSSION TITLE
Unable to connect with SignalR with any available transports. Error 404 "No Connection with that ID"

### DISCUSSION THREAD

**Discussion by mustafajavid** [ThreeMammals/Ocelot#1985] [Show and tell]
## Problem
I have a react website in which I am trying to connect to SignalR hub on my .net core server via ocelot gateway.

I am unable to connect to hub and get a 404 error "No Connection with that ID". I have only one server instance running and no load balancer is involved.
![SignalR  Hub request](https://user-images.githubusercontent.com/35831655/104721101-2f263c80-574f-11eb-87e8-d67813ac25ed.png)
![Hub Call response](https://user-images.githubusercontent.com/35831655/104721135-3b11fe80-574f-11eb-87b1-df43b330b015.png)

Even though the negotiate call is always successful. 
![Negotiate Response](https://user-images.githubusercontent.com/35831655/104722709-11a5a280-5750-11eb-8942-f42591c4ae51.png)

WebSocket and Server Sent Events are not working at all and on Long Polling, 404 is returned.
![Console logs](https://user-images.githubusercontent.com/35831655/104723199-b7f1a800-5750-11eb-9f71-5c3b1e3919a0.png)

Network call logs are:
![Network logs](https://user-images.githubusercontent.com/35831655/104723286-dd7eb180-5750-11eb-9f0c-c17a90a0a1e7.png)

I have added all possible routes for ocelot. I am out of ideas on what to do now.
![SignalR Ocelot Routes](https://user-images.githubusercontent.com/35831655/104723486-1fa7f300-5751-11eb-9509-35cd5ceedb0f.png)

I have observed that the connectionId returned in response to the negotiate call is not used to connect to hub and instead the connectionToken value is used as the value for "id" query string parameter.

## Specifications

  - Ocelot Version: 15.0.6
  - @microsoft/signalr Version: 5.0.1
  - .Net Core: 3.1
  - Platform: Windows

**Comment by joaopgrassi** [comment]
Not sure this is going to help you, but in our setup we have this:

SignalR service:
```
// Startup.cs > Configure
app.UseEndpoints(endpoints =>
{
    endpoints.MapHub<AppHub>("notify");
});
```

Then in ocelot, we have this route:
```
{
  "DownstreamPathTemplate": "/{everything}",
  "DownstreamScheme": "ws",
  "UpstreamPathTemplate": "/my-service/notifications/{everything}",
  "DownstreamHostAndPorts": [
  	{
  		"Host": "localhost",
  		"Port": 5036
  	}
  ],
  "UpstreamHttpMethod": [ "GET", "POST", "PUT", "DELETE", "OPTIONS" ],
},
```

Then, we connect from our front-end app like this (in typescript..)

```
import * as signalR from '@microsoft/signalr';

this.hubConnection = new signalR.HubConnectionBuilder()
	.configureLogging(signalR.LogLevel.Information)
	.withUrl(".../my-service/notifications/notify")
	.withAutomaticReconnect([0, 3000, 5000, 10000, 15000, 30000])
	.build();
```

And about this:

>I have observed that the connectionId returned in response to the negotiate call is not used to connect to hub and instead the connectionToken value is used as the value for "id" query string parameter.

AFAIK that is how it should be. If you, for example, look at the sample app here
https://docs.microsoft.com/en-us/aspnet/core/tutorials/signalr-typescript-webpack?view=aspnetcore-5.0&tabs=visual-studio

The value sent on `hub?id=` is the `connectionToken` returned from the negotiate call. It seems it's some error with your re-routes really.

Also, make sure you don't have any other routes that are selected first. Ocelot looks at them in order they are defined, so if a same route is defined first that is going to fail.

Another thing, maybe to check is when you run your SignalR service, do you have ports for http and https? Are those correctly configured in the routes in Ocelot?

And finally, remember to add `app.UseWebSockets()` in your Ocelot service: https://ocelot.readthedocs.io/en/latest/features/websockets.html#signalr

**Reply by raman-m** [reply]
Great instructions! Thanks!

**Reply by raman-m** [reply]
### MS Learn | [ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/)
- [Tutorial: Get started with ASP.NET Core SignalR using TypeScript and Webpack](https://learn.microsoft.com/en-us/aspnet/core/tutorials/signalr-typescript-webpack?view=aspnetcore-8.0&tabs=visual-studio)

**Comment by Kation** [comment]
Ccheck your signalr response time with `/hub/negotiate`
Make sure it response less than 5 secs.

**Comment by quoctienkt** [comment]
I have the same issue, but it should be issue of signalR-self (SevertSentEvent transport issue)

**Comment by Magnona** [comment]
i have the same issue, did anyone found solution for this error?

**Comment by quoctienkt** [comment]
> i have the same issue, did anyone found solution for this error?

I figured it out
- browser limits concurrent connection to lower than 6, so you will need to: use websocket for signalr transportation or enable http/2 or http/3 from your application and hosting

**Reply by raman-m** [reply]
Interesting...
Are you expert with SignalR?

**Comment by mhmoudalaskalany** [comment]
i set up most of the configuration here and i was receiving error as 1006
i figured out that i am missing to add  CONNECT in the up stream method , seems the signalr package in angular is using this verb

These are my config in the gateway routes
```json
{
   "UpstreamHttpMethod": [ "GET", "POST", "PUT", "DELETE", "OPTIONS", "CONNECT" ],
   "UpstreamPathTemplate": "/ptmapi/hubs/{catchAll}",
   "DownstreamPathTemplate": "/hubs/{catchAll}",
   "DownstreamScheme": "wss",
   "DownstreamHostAndPorts": [
     {
       "Host": "localhost",
       "Port": 5026
     }
   ],
   "DangerousAcceptAnyServerCertificateValidator": true
}
```

**Reply by raman-m** [reply]
Thanks for your user scenario!
